### PR TITLE
Add CSDS validation ability to PSM Sec framework

### PR DESCRIPTION
**Pull request by lidizheng**
As title.

The CSDS validation is part of baseline test now. It can be simply extended for other test cases.

I can't get the firewall exemption, so I added the functionality to create/delete firewall. It works pretty stably.

b/187885884

### CODE REVIEW

**Review by sergiitk — CHANGES_REQUESTED** [review]
I doubt whether this would make sense to add firewall rules as part of test runs. Let's discuss it IRL.

**Comment by lidizheng** [comment]
@sergiitk PTALA.

**Comment by lidizheng** [comment]
@sergiitk Is there anything else I need to do? The firewall-creation is moved to https://github.com/grpc/grpc/pull/26255.

File: tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py [hunk 1]
```diff
@@ -155,6 +156,20 @@ def assertSuccessfulRpcs(self,
...
+        seen = set()
+        want = frozenset([
+            'listener_config', 'cluster_config', 'route_config',
+            'endpoint_config'
+        ])
+        for xds_config in config.xds_config:
+            seen.add(xds_config.WhichOneof('per_xds_config'))
+        logger.debug('Received xDS config dump: %s',
+                     json_format.MessageToJson(config, indent=2))
+        self.assertEqual(want, seen)
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py:171]:
```suggestion
        self.assertSameElements(want, seen)
```

This one has a bit nicer output. F.e `assertSameElements()` output if `endpoint_config` missing, and `unexpected_config` present:

```
AssertionError: Expected, but missing:
  ['endpoint_config']
Unexpected, but present:
  ['unexpected_config']
```

Vs `assertEqual()`:
```
AssertionError: {'endpoint_config', 'listener_config', 'c[26 chars]fig'} != frozenset({'listener_config', 'unexpected[39 chars]ig'})
```

**Review by sergiitk — APPROVED** [review]
Looks great, minor comments/improvements

File: tools/run_tests/xds_k8s_test_driver/framework/rpc/grpc_csds.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import grpc
+from envoy.service.status.v3 import csds_pb2
+from envoy.service.status.v3 import csds_pb2_grpc
+
+# Envoy protos provided by PyPI package xds-protos
+# Needs to import the generated Python file to load descriptors
+from envoy.extensions.filters.network.http_connection_manager.v3 import http_connection_manager_pb2
+from envoy.extensions.filters.common.fault.v3 import fault_pb2
+from envoy.extensions.filters.http.fault.v3 import fault_pb2
+from envoy.extensions.filters.http.router.v3 import router_pb2
```
[the remark this answers is not in the document]
Comment (lidizheng) [reply tools/run_tests/xds_k8s_test_driver/framework/rpc/grpc_csds.py:32]: `csds.proto` transitively imports `Cluster` and `Listener` and `NodeMatcher`. Other ones are the same.

File: tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py [hunk 3]
```diff
@@ -155,6 +156,20 @@ def assertSuccessfulRpcs(self,
             0,
             msg=f'Expected all RPCs to succeed: {failed} of {num_rpcs} failed')
 
+    def assertXdsConfigExists(self, test_client: XdsTestClient):
+        config = test_client.csds.fetch_client_status()
```
[the remark this answers is not in the document]
Comment (lidizheng) [reply tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py]: Sounds good. I didn't notice there is a logging functionality here.

File: tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py (hunk 1 above)
Comment (lidizheng) [reply tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py]: TIL. Good idea!

File: tools/run_tests/xds_k8s_test_driver/tests/baseline_test.py [hunk 4]
```diff
@@ -53,7 +53,10 @@ def test_traffic_director_grpc_setup(self):
         with self.subTest('7_start_test_client'):
             test_client: _XdsTestClient = self.startTestClient(test_server)
 
-        with self.subTest('8_test_server_received_rpcs_from_test_client'):
+        with self.subTest('8_validate_xDS_config'):
```
[the remark this answers is not in the document]
Comment (lidizheng) [reply tools/run_tests/xds_k8s_test_driver/tests/baseline_test.py]: Name changed.

**Review by sergiitk — CHANGES_REQUESTED** [review]
This is blocked at the moment, confirmed one of interop tests doesn't have CSDS enabled. Sorry.

File: tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py [hunk 5]
```diff
@@ -155,6 +156,22 @@ def assertSuccessfulRpcs(self,
...
 
+    def assertXdsConfigExists(self, test_client: XdsTestClient):
+        config = test_client.csds.fetch_client_status(log_level=logging.INFO)
+        self.assertIsNotNone(config)
+        seen = set()
+        want = frozenset([
+            'listener_config',
+            'cluster_config',
+            'route_config',
+            'endpoint_config'
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py:167]:
```suggestion
            'endpoint_config',
```
The comma got lost.

**Review by sergiitk — APPROVED** [review]
This is unblocked now.

File: tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py (hunk 5 above)
Comment (lidizheng) [reply tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py]: > The presence of a trailing comma is also used as **a hint** to our Python code auto-formatter YAPF to direct it to auto-format the container of items to **one item per line** when the , after the final element is present.

If the trailing-comma is present, the behavior of `yapf` is okay. But if the trailing-comma is missing, I think `yapf` should be more smart, whether to add the comma and stretch the list, or move the brackets to the ending line.

**Pull request merged**